src/backends/backend/onnxruntime/onnxruntime.cc

```diff
@@ -0,0 +1,1873 @@
+
+  GraphOptimizationLevel optimization_level =
+      GraphOptimizationLevel::ORT_ENABLE_ALL;
+  {
+    ni::TritonJson::Value optimization;
+    if (ModelConfig().Find("optimization", &optimization)) {
+      ni::TritonJson::Value graph;
+      if (optimization.Find("graph", &graph)) {
+        int64_t graph_level = 0;
+        THROW_IF_BACKEND_MODEL_ERROR(graph.MemberAsInt("level", &graph_level));
+        if (graph_level == -1) {
+          optimization_level = GraphOptimizationLevel::ORT_ENABLE_BASIC;
+        } else if (graph_level == 1) {
+          optimization_level = GraphOptimizationLevel::ORT_ENABLE_EXTENDED;
+        }
+      }
+    }
+  }
+  THROW_IF_BACKEND_MODEL_ORT_ERROR(
+      ort_api->SetSessionGraphOptimizationLevel(soptions, optimization_level));
+
+  // FIXME. Is it possible to share a single OrtSession across
+  // multiple instances? If so then should move loading and validation
+  // of the session to here instead of creating a session for each
+  // instance in ModelStateInstance::Create().
+}
+
+TRITONSERVER_Error*
+ModelState::LoadModel(
+    const std::string& artifact_name,
+    const TRITONSERVER_InstanceGroupKind instance_group_kind,
+    const int32_t instance_group_device_id, std::string* model_path,
+    OrtSession** session, OrtAllocator** allocator)
+{
+  // Find the ONNX file that describes the model itself. If the model
+  // configuration doesn't have an explicit model file specified then
+  // use the default name ("model.onnx").
+  std::string cc_model_filename = artifact_name;
+  if (cc_model_filename.empty()) {
+    cc_model_filename = "model.onnx";
+  }
+
+  *model_path = nib::JoinPath(
+      {RepositoryPath(), std::to_string(Version()), cc_model_filename});
+
+  // If the model path is a directory then the actual model is
+  // <dir>/model.onnx.
+  {
+    bool is_dir;
+    RETURN_IF_ERROR(nib::IsDirectory(*model_path, &is_dir));
+    if (is_dir) {
+      *model_path = nib::JoinPath({*model_path, "model.onnx"});
+    }
+  }
+
+  {
+    bool exists;
+    RETURN_IF_ERROR(nib::FileExists(*model_path, &exists));
+    RETURN_ERROR_IF_FALSE(
+        exists, TRITONSERVER_ERROR_UNAVAILABLE,
+        std::string("unable to find '") + *model_path +
+            "' for model instance '" + Name() + "'");
+  }
+
+  // Make a clone for the session options for this instance...
+  OrtSessionOptions* soptions;
+  RETURN_IF_ORT_ERROR(
+      ort_api->CloneSessionOptions(session_options_.get(), &soptions));
+  std::unique_ptr<OrtSessionOptions, SessionOptionsDeleter> soptions_wrapper(
+      soptions);
+
+  bool need_lock = false;
+
+  // Execution providers if they are requested...
+  if (instance_group_kind != TRITONSERVER_INSTANCEGROUPKIND_AUTO) {
+    // Default GPU execution provider.
+#ifdef TRITON_ENABLE_GPU
+    if (instance_group_kind == TRITONSERVER_INSTANCEGROUPKIND_GPU) {
+      RETURN_IF_ORT_ERROR(OrtSessionOptionsAppendExecutionProvider_CUDA(
```

Comment:
The order of calling `OrtSessionOptionsAppendExecutionProvider_XXX()` implies the priority. CUDA should be moved after checking GPU execution accelerator as "default" GPU provider

Reply:
Fixed